### PR TITLE
Stop modifying classes and always create new ones

### PR DESCRIPTION
**Pull request by hynek**
This allows us to harmonize class creation for dict and slots classes.

Interestingly this seems to break our Hypothesis strategies.  I can't dig any further right now but I wanted to put it out there.  Helpful comments welcome.

Ref #223 #202

### CODE REVIEW

**Comment by Tinche** [comment]
I’m happy to take a look

**Comment by hynek** [comment]
A few more notes:

1. I'm not married to the naming.
2. I’m not raising a warning because it seems unfortunate to force people into setting the argument and makes it kind of impossible to remove it later.  My hope is that most people don't ever realize that we changed something underneath.
3. I hate Python 2.

**Comment by Tinche** [comment]
Hm after a 5 minute investigation, there seems to be a problem with `copy.copy` on classes that are `replace_class=True, frozen=True` (and only those, slot classes are OK). Will look into it more later.

On a side note, is it possible to overwrite the class `__repr__`? As in, not the instance repr. Our Hypothesis classes are called `HypClass`, would be nicer if their repr had the flags too. I mean, I could always just change the name...

**Comment by hynek** [comment]
🤦‍♂️

File: changelog.d/260.deprecation.rst [hunk 1]
```diff
@@ -0,0 +1,8 @@
+Added new option to ``@attr.s()``: *replace_class* will make ``attrs`` stop attaching methods to the existing class.
+Instead it will create a new class with the methods attached.
+This is consistent with how ``slots``-classes have *always* been created.
+
+For the next year, the default value is ``replace_classes=False`` which is the current behavior.
+
+The default value will change to ``True`` no sooner than one year after this release.
+There are no concrete plans to remove ``replace_classes=False`` altogether but its use is discouraged.
```
Comment (Tinche) [inline changelog.d/260.deprecation.rst]: After a year, if no problems present themselves, I say we remove the option altogether. Mainly in the interest of simplicity, `attr.s` seems to be getting options all the time.

Comment (hynek) [reply changelog.d/260.deprecation.rst]: The problem is that we’d break everyone’s software who decided to do the right thing and set it to true.

Comment (hynek) [reply changelog.d/260.deprecation.rst]: I guess what we could do is to add a `**kw` and spit out DeprecationWarnings on certain arguments that have been removed.  But I really don’t want `attrs` to become notorious for breaking backward-comparability all the time.  We’re already getting comments like https://www.reddit.com/r/Python/comments/74fro5/attrs_iii_frozen_classes/dnzjn6d/ :|

Comment (Tinche) [reply changelog.d/260.deprecation.rst]: Why don't we just do this: default to False for one year, default to True for one year, remove if no problems. I mean, the alternative is to carry this forever :/

I'm ok either way btw, this is definitely not a hill I want to die on :) I'm just worried about attr.s having a gazillion args down the line, which will make it harder to use.

Comment (hynek) [reply changelog.d/260.deprecation.rst]: Yeah well I’m deeply unhappy with every option, hence trying to source opinions. :)  The two years plan is something I considered too.  Let's see we can get some opinions out of glyph and Nate.

**Comment by hynek** [comment]
So um @glyph and @njsmith, any ideas/opinions on how to make this thing happen without pissing people off unnecessarily?

**Comment by hynek** [comment]
Just to add another thought: my wish is that *ideally*, nobody notices a difference at all.  So a three stage approach with a `*kw`-warning in two years seems best:

1. now: add option, default is False
2. Oct 2018: default is True
3. Oct 2019: option becomes a nop and starts raising a warning

Now, this kind of costs us a pointless year because people are not gonna try out that option unless they have a problem.  It's like with pre-releases, nobody tests them. 

Ergo, in Oct 2018, there will be people yelling at us in any case and we've maybe saved <0.01% of our users some headaches.

***

OTOH, I think I’ll do it anyway because I'm totally not ready to break hashing a second time this year. :|  Also we'll be able to point to this: yes we did the right thing.

I really hope this is the last fundamental screwup in `attrs`. :|

***

I’d really like more input on this…happily also from @wsanchez @Lukasa @DRMacIver @wbolster @hawkowl

**Comment by hynek** [comment]
I have added the current plan to the news fragment.

**Comment by DRMacIver** [comment]
I don't think I actually understand the problem well enough to comment yet. What actually is the interaction with hashing and equality here that causes the problem?

I assume the "super" problem is that if I have something like:

```
@attr.s()
class Foo(Bar):
    def stuff(self):
        return super(Foo, self).stuff()
```

Then the "Foo" in the method is different from the Foo outside the method and this causes a supertype mismatch of some sort? I assume the class you're creating *can't* be a subclass of the existing Foo. Are you copying over the namespace and is that what causes you problems?

**Comment by DRMacIver** [comment]
OK. I've now done the reading of the news fragment and understand some of this better.

On the hashing thing... I think I'd personally be inclined to be less nice than you. I would argue that getting a hash when passing `hash=False` is just a bug and people relying on it are relying on a bug, and that just fixing that bug would be a non-breaking change. As you say, people are going to suffer that change at some point anyway, so I think you might as well just rip off the band-aid.

Personally the idea of adding a `replace_class` argument to `attr.s` doesn't sit well with me, especially when defaulting it to false. One of the rules we have in Hypothesis code review is that when adding a setting you're supposed to ask whether the user is ever likely to actually have a reason to care about the value of this setting. When adding a new scary change the user *might* care about being able to back it out, but they're certainly not going to care to opt into it because `replace_class` is about as internal to attrs as it gets.

Were it me I would adopt the following plan:

* No optional behaviour. Just opt everyone into this straight up.
* Think real hard about possible gotchas and how to fix them.
* Before releasing, try running the test suite of a few of the larger open source users of attrs and see if anything hilarious happens.
* Be on the ball for a certain amount of user screaming.
* In the unlikely event that there are any actual unfixable bugs caused by this, add a `replace_class` option defaulting to True that emits a warning when you set it to False.

Note however this is coming from the guy who thought "Lets get real drunk and turn coverage on for every Hypothesis user" was a good idea, so I may have more of a YOLO attitude to feature evolution than you are comfortable with. 😉

**Comment by wsanchez** [comment]
I tend to agree with @DRMacIver, though it's possible that all the wine I had over the weekend is influencing me into associating with the "Lets get real drunk and…" camp.

I certainly agree regarding the hashing behavior; that's a bug and it should just get fixed.  I don't see a strong argument for bending over to provide compatibility there; at a minimum, I'd have `replace_class` default to `True`, but as a user I'd rather fix the bug if this effects me than opt into buggy behavior.  (But, to be fair, I also don't expect his to effect me.)

**Comment by hynek** [comment]
I’d like to stress one thing: `hash=False` is supposed to give you hash-by-id, if**f**, `cmp=False` on Python 3.  It will *always* give you one on Python 2.

What we *could* do is using this as a unification/simplification for creating classes and attach `object.__hash__` for non-slot classes to not breaking any compatibility.

***

Then continue with #202 and use an enum to control what actually happens and deprecate using bool for the hash argument. 🤔

**Comment by DRMacIver** [comment]
@hynek that sounds sensible to me.

**Comment by hynek** [comment]
OK, I’ve spent yesterday basically rewriting the class generation code because the old one was all about tacking on and then copying or removing.

Now it's much cleaner.  There might be some doc inconsistencies I guess?  I’m much happier with how it works now.

File: changelog.d/223.change.rst [hunk 2]
```diff
@@ -0,0 +1,4 @@
+``attrs``-decoratorated classes are never modified now.
+Instead a new class is created with the appropriated methods and attributes attached from the get go.
+
+For backward-compatibility, the behavior of getting hashing by ID on Python 3 when passing ``@attr.s(hash=False)`` is manually retained by setting ``__hash__`` to ``object.__hash__``.
```
Comment (wsanchez) [inline changelog.d/223.change.rst]: Something is grammatically off here… maybe:

For backward-compatibility, the prior behavior of hashing by ID on Python 3 when passing ``@attr.s(hash=False)`` may be retained retained by setting ``__hash__`` to ``object.__hash__``.

File: changelog.d/260.change.rst [hunk 3]
```diff
@@ -0,0 +1,4 @@
+``attrs``-decoratorated classes are never modified now.
+Instead a new class is created with the appropriated methods and attributes attached from the get go.
+
+For backward-compatibility, the behavior of getting hashing by ID on Python 3 when passing ``@attr.s(hash=False)`` is manually retained by setting ``__hash__`` to ``object.__hash__``.
```
Comment (wsanchez) [inline changelog.d/260.change.rst]: See above.  Also, I'm confused about why there are two of these, but I bet there's a reason.

**Comment by DRMacIver** [comment]
> What we could do is using this as a unification/simplification for creating classes and attach object.__hash__ for non-slot classes to not breaking any compatibility.

Sorry, I somehow misparsed what you originally said which was totally unambiguous but not quite what I thought the good idea was.

What I would actually recommend here is that the default behaviour is to attach a `__hash__` which calls `object.__hash__` and emits a deprecation warning saying that if this is the behaviour you want then you should be explicit about wanting it.

**Comment by hynek** [comment]
> What I would actually recommend here is that the default behaviour is to attach a __hash__ which calls object.__hash__ and returns its value, but also emits a deprecation warning saying that if this is the behaviour you want then you should be explicit about wanting it.

The problem is that as of now, there is no explicit way of defining it.  You totally can ask for unhashable but with cmp and the behavior also differs between Python versions.  That’s why I want to start raising a warning as soon as there is actually a good way.

File: changelog.d/260.change.rst (hunk 3 above)
Comment (hynek) [reply changelog.d/260.change.rst]: Yeah there is: identical news fragments are merged into one changelog entry by town crier with multiple links to the issues.

File: changelog.d/223.change.rst (hunk 2 above)
Comment (hynek) [reply changelog.d/223.change.rst]: Oh dear, so my wording was so bad that it confused you what I mean. 🙈

*We* do set `__hash__` to `object.__hash__` automatically, to not break compat.

English is hard man.  Let me try again…

> The previous (incorrect) behavior of hashing by object ID on Python 3 when passing `@attr.s(hash=False)` (which implies `cmp=True`), is retained  for backward compatibility reasons.
>
> A better -- and more explicit -- way to control object hashing is in the pipeline.

Better/clearer?

Comment (wsanchez) [reply changelog.d/223.change.rst]: Ah, yes, OK.  :-)

**Comment by hynek** [comment]
OK folks, I get it: this PR is fucking scary.

So I’m not gonna ask you to review and risk to share the blame.  Mut what I *am* gonna ask you is to run it against your code bases and report back any breakage you encounter.  Technically, nothing should break.

I just did that and the result is 6dd4b1d.  I’ll take all the blame, but please help me not have an emergency 17.4 by running your test suite once. :D

File: src/attr/_make.py [hunk 4]
```diff
@@ -181,22 +181,24 @@ class MyClassAttributes(tuple):
...
-    Return a list of tuples of (attribute name, attribute).
+    Return an `_Attributes`.
     """
     if these is None:
         ca_list = [(name, attr)
                    for name, attr
                    in cls.__dict__.items()
                    if isinstance(attr, _CountingAttr)]
-        for name, _ in ca_list:
-            delattr(cls, name)
```
Comment (DRMacIver) [inline src/attr/_make.py]: To check my understanding: This was only needed because the old class was still hanging around, and now it's not because they're just not copied to the newly created one anyway class?

**Comment by DRMacIver** [comment]
FWIW I've run the Hypothesis test suite with these changes and it seems fine, but I also don't think Hypothesis makes interesting enough use of attrs that that's a meaningful test of anything.

**Comment by Tinche** [comment]
I got one:
```
import attr
from typing import TypeVar, Generic

G = TypeVar('G', bound=int)


@attr.s
class A(Generic[G]):
    a = attr.ib()
```

Works with PyPI, not with this branch (on 3.6). GL :D

File: src/attr/_make.py [hunk 5]
```diff
@@ -263,6 +272,161 @@ def _frozen_delattrs(self, name):
...
+        # compiler will bake a reference to the class in the method itself
+        # as `method.__closure__`.  Since we replace the class with a
+        # clone, we rewrite these references so it keeps working.
+        for item in cls.__dict__.values():
+            closure_cells = getattr(item, "__closure__", None)
+            if not closure_cells:  # Catch None or the empty list.
+                continue
+            for cell in closure_cells:
+                if cell.cell_contents is self._cls:
+                    set_closure_cell(cell, cls)
```
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py]: It is, isn't it?  I believe @Tinche sold his soul for it. :D

File: src/attr/_make.py [hunk 6]
```diff
@@ -206,35 +208,42 @@ def _transform_attrs(cls, these):
...
     for c in reversed(cls.__mro__[1:-1]):
         sub_attrs = getattr(c, "__attrs_attrs__", None)
         if sub_attrs is not None:
-            super_cls.extend(
-                a for a in sub_attrs
-                if a not in super_cls and a.name not in non_super_names
+            super_attrs.extend(
+                a
+                for a in sub_attrs
+                if a not in super_attrs and a.name not in non_super_names
```
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py]: Well, we need to add each only once and we can’t use a set, because the default might not be hashable. :|

Suggestions/recommendations?

File: src/attr/_make.py (hunk 4 above)
Comment (hynek) [reply src/attr/_make.py]: Yeah, the original approach was to stick everything on the old class and if it’s a slots class, copy the result to a new one.

**Comment by hynek** [comment]
Thanks @DRMacIver, your look is very much appreciated!

@Tinche not sure if trolling? :)

File: src/attr/_make.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DRMacIver) [reply src/attr/_make.py]: Well ignoring the O(n^2) thing, I think an explicit loop would be *much* clearer here:

```python
for a in sub_attrs:
    if a not in super_attrs and a.name not in non_super_names:
        super_attrs.append(a)
```

That way you at least don't have to deal with the question of how iteration and updating the list interact (which is the bit that confused me).

In terms of suggesting how to improve this complexity wise, I don't think it matters much (there aren't *that* many attrs and it only runs once), but I don't think I understand the code well enough. Why are there duplicates in `sub_attrs` at all?

**Comment by Tinche** [comment]
I'm not trolling, this was a failure from my work codebase. I've simplified it of course.

**Comment by hynek** [comment]
Related to the type work, not this PR right?  Maybe open new issue and set milestone?

**Comment by wsanchez** [comment]
Well so I was wrong… this did break my code, specifically, [this test](https://github.com/burningmantech/ranger-ims-server/blob/master/src/ims/model/test/test_cmp.py#L49).

This isn't my favorite code, so please don't judge me.  :-)

I have a hack mix-in class I use for attrs-classes where I want some simple control over how comparisons are done… anyway it fails [here](https://github.com/burningmantech/ranger-ims-server/blob/master/src/ims/model/test/test_cmp.py#L55).

**Comment by wsanchez** [comment]
(I'm 100% OK with changing this code, by the way, just reporting on test results, as requested.)

File: src/attr/_make.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py]: > Why are there duplicates in sub_attrs at all?

I think something something inheritance?  I think I’ll do the loop thing and investigate, thanks!  Your “naive” questions are *very* helpful.

File: src/attr/_make.py [hunk 7]
```diff
@@ -347,75 +511,39 @@ def attrs(maybe_cls=None, these=None, repr_ns=None,
...
+            builder.add_cmp()
 
         if hash is not True and hash is not False and hash is not None:
+            # Can't use `hash in` because 1 == True for example.
             raise TypeError(
                 "Invalid value for hash.  Must be True, False, or None."
             )
         elif hash is False or (hash is None and cmp is False):
-            pass
+            builder.add_hash_by_id()
```
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py]: No, because the whole point of this patch is that dict classes will behave like slots classes now.  A comment should be added tho indeed!  (your question genuinely confused me and made me think it through one more time, so thanks again for that!)

File: src/attr/_make.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (hynek) [reply src/attr/_make.py]: Turns out™, you’ve spotted a bug!  Check out 507d9d5baae3871ff0392cf7794165b45c1e6fac for details if you want to know more.

Does it make more sense now?

**Comment by hynek** [comment]
Great news everyone, we’ve got ourselves another edge case:

```python
import attr

@attr.s
class BaseNoSlots:
  def __init_subclass__(cls, param, **kwargs):
    super().__init_subclass__(**kwargs)
    cls.param = param

@attr.s    
class Foo(BaseNoSlots, param='foo'):
  pass

print(Foo.param)
```

This is taken from #269 and https://repl.it/Mmkb/0

Since we didn’t intervene into the class generation for non-slots before, it just worked.

Now we get:

```pycon
>>> import attr
...
... @attr.s
... class BaseNoSlots:
...   def __init_subclass__(cls, param, **kwargs):
...     super().__init_subclass__(**kwargs)
...     cls.param = param
...
... @attr.s
... class Foo(BaseNoSlots, param='foo'):
...   pass
Traceback (most recent call last):
  File "<stdin>", line 10, in <module>
  File "/Users/hynek/Projects/attrs/src/attr/_make.py", line 573, in attrs
    return wrap(maybe_cls)
  File "/Users/hynek/Projects/attrs/src/attr/_make.py", line 566, in wrap
    return builder.build()
  File "/Users/hynek/Projects/attrs/src/attr/_make.py", line 379, in build
    cd,
TypeError: __init_subclass__() missing 1 required positional argument: 'param'
__init_subclass__() missing 1 required positional argument: 'param'
```

**Comment by hynek** [comment]
@RonnyPfannschmidt sooo…how could metaclasses help here without making stuff runtime-dynamic?

**Comment by hynek** [comment]
In any case: I’m going to make this PR only about the refactoring which is a net win.  Then we’ll get attrs 17.3 out of the door and then we‘ll tackle it again.  Maybe we’ll find a better way to attach __slots__.
